aten/src/ATen/native/cuda/Normalization.cuh

```diff
@@ -857,4 +957,690 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda_template(
+  int c_offset = blockIdx.x * blockDim.x + threadIdx.x;
+
+  auto m_c = mean[c_offset];
+  auto m_dy_c = sum_dy[c_offset] / div;
+  auto factor_1_c = inv_std[c_offset];
+  auto factor_2_c = (weight == nullptr? accscalar_t(1.0) : static_cast<accscalar_t>(weight[c_offset])) * factor_1_c;
+  factor_1_c = factor_1_c * factor_1_c * sum_dy_xmu[c_offset] / div;
+
+  int loop_count = 1 + (reduction_size - 1) / (inner_loop_stride * PARALLEL_LOADS);
+  int address_base = m_offset * stride + c_offset;
+  int address_increment = inner_loop_stride * stride;
+
+  for (int i = 0; i < loop_count; i++) {
+#pragma unroll
+    for (int j = 0; j < PARALLEL_LOADS; j++) {
+      if (c_offset < stride && m_offset < reduction_size) {
+        grad_input[address_base] = static_cast<scalar_t>(
+            (static_cast<accscalar_t>(grad_output[address_base]) - m_dy_c -
+            (static_cast<accscalar_t>(input[address_base]) - m_c) * factor_1_c)
+            * factor_2_c);
+      }
+      m_offset += inner_loop_stride;
+      address_base += address_increment;
+    }
+  }
+}
+
+template<typename scalar_t>
+std::tuple<Tensor, Tensor> batch_norm_stats_channels_last_cuda_template(const Tensor& input, double epsilon) {
+  using accscalar_t = at::acc_type<scalar_t, true>;
+
+  const auto stride = input.size(1);
+  const auto reduction_size = input.numel() / stride;
+
+  auto scalar_type = input.scalar_type() == at::kHalf ? at::kFloat : input.scalar_type();
+  auto option = input.options().dtype(scalar_type);
+
+  at::Tensor out_invstd = at::empty({stride}, option);
+  at::Tensor out_mean = at::empty({stride}, option);
+
+  dim3 block;
+  dim3 grid;
+  flexible_launch_configs(reduction_size, stride, block, grid, true);
+
+  at::Tensor staging_data;
+  at::Tensor semaphores;
+  if (grid.y > 1) {
+    staging_data = at::empty({4*stride*grid.y}, option);
+    semaphores = at::zeros({grid.x}, input.options().dtype(at::kInt));
+  }
+
+  accscalar_t* staging_data_ptr = grid.y > 1 ? staging_data.data_ptr<accscalar_t>() : nullptr;
+  int* semaphores_ptr = grid.y > 1 ? semaphores.data_ptr<int>() : nullptr;
+  batch_norm_collect_statistics_channels_last_kernel<InvStd, scalar_t, accscalar_t, ELEMENTS_PER_ITER>
+      <<<grid, block, 0, at::cuda::getCurrentCUDAStream()>>>(
+      input.data_ptr<scalar_t>(),
+      out_mean.data_ptr<accscalar_t>(),
+      out_invstd.data_ptr<accscalar_t>(),
+      staging_data_ptr,
+      semaphores_ptr,
+      reduction_size,
+      stride,
+      epsilon);
+  AT_CUDA_CHECK(cudaGetLastError());
+
+  return std::make_tuple(out_mean, out_invstd);
+}
+
+void batch_norm_elemt_channels_last_cuda_template(
+    at::Tensor& output,
+    const at::Tensor& input,
+    const at::Tensor& weight,
+    const at::Tensor& shift,  // bias of BN
+    const at::Tensor& mean,
+    const at::Tensor& inv_std,
+    double epsilon,
+    const at::optional<at::Tensor>& z,  // bias after BN
+    const bool fuse_relu) {
+  const auto stride = input.size(1);
```

Comment:
you probably need to check that input is at least 2d. After than input.sizes()[1] is slightly faster

Reply:
It is checked in python https://github.com/pytorch/pytorch/blob/bf4fcab68104b333c81e39d142dccc7ae7a6f1d7/torch/nn/modules/batchnorm.py#L474-L477